drivers/sdhc/sdhc_max32.c

```diff
@@ -0,0 +1,412 @@
+	return 0;
+}
+
+static int sdhc_max32_card_busy(const struct device *dev)
+{
+	int ret = 0;
+
+	ret = MXC_SDHC_Card_Busy();
+
+	return ret;
+}
+
+static int sdhc_max32_reset(const struct device *dev)
+{
+	MXC_SDHC_Reset();
+
+	return 0;
+}
+
+static int sdhc_max32_request(const struct device *dev, struct sdhc_command *cmd,
+			      struct sdhc_data *data)
+{
+	int ret = 0;
+	unsigned int mxc_cmd = 0;
+	mxc_sdhc_cmd_cfg_t sd_cmd_cfg;
+	bool card_size_workaround = false;
+
+	if (data) {
+		sd_cmd_cfg.sdma = (unsigned int)data->data;
+		sd_cmd_cfg.block_size = data->block_size;
+		sd_cmd_cfg.block_count = data->blocks;
+	}
+
+	sd_cmd_cfg.arg_1 = cmd->arg;
+	sd_cmd_cfg.dma = true; /* todo: add config depending on config_dma etc. */
+
+	switch (cmd->opcode) {
+	case SD_READ_SINGLE_BLOCK:
+	case SD_READ_MULTIPLE_BLOCK:
+		sd_cmd_cfg.direction = MXC_SDHC_DIRECTION_READ;
+		sd_cmd_cfg.arg_1 = data->block_addr;
+		break;
+	case SD_WRITE_SINGLE_BLOCK:
+	case SD_WRITE_MULTIPLE_BLOCK:
+		sd_cmd_cfg.direction = MXC_SDHC_DIRECTION_WRITE;
+		sd_cmd_cfg.arg_1 = data->block_addr;
+		break;
+	case SD_SEND_CSD:
+		card_size_workaround = true;
+	default:
+		sd_cmd_cfg.direction = MXC_SDHC_DIRECTION_CFG;
+		break;
+	}
+
+	ret = cmd_opcode_converter(cmd->opcode, &mxc_cmd);
+	if (ret) {
+		return ret;
+	}
+	sd_cmd_cfg.command = mxc_cmd;
+	sd_cmd_cfg.host_control_1 = MXC_SDHC_Get_Host_Cn_1();
+	sd_cmd_cfg.callback = NULL;
+
+	/*
+	 * todo: this was also needed, otherwise applications failed randomly. it would be good to
+	 * remove this with a better solution in future.
+	 */
+	k_sleep(K_MSEC(1));
+	ret = MXC_SDHC_SendCommand(&sd_cmd_cfg);
+	if (ret) {
+		LOG_ERR("MXC_SDHC_SendCommand error:%d, SD opcode: %d", ret, cmd->opcode);
+		return ret;
+	}
+
+	MXC_SDHC_Get_Response128((char *)(cmd->response));
+
+	if (card_size_workaround) {
+		/*
+		 * this workaround is required for only CMD9. This fixes size problem. Otherwise
+		 * it doesn't give the correct device size information.
+		 */
+		cmd->response[1] <<= 8;
+		cmd->response[3] <<= 8;
+	}
+
+	return 0;
+}
+
+static int sdhc_max32_get_card_present(const struct device *dev)
+{
+	return MXC_SDHC_Card_Inserted();
+}
+
+static int sdhc_max32_get_host_props(const struct device *dev, struct sdhc_host_props *props)
+{
+	struct sdhc_max32_data *sdhc_data = dev->data;
+
+	memcpy(props, &sdhc_data->props, sizeof(struct sdhc_host_props));
+
+	return 0;
+}
+
+static int sdhc_max32_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	struct sdhc_max32_data *data = dev->data;
+	struct sdhc_host_props *props = &data->props;
+	enum sdhc_clock_speed speed = ios->clock;
+	unsigned int clk_div = 0;
+
+	if (speed) {
+		if (speed < props->f_min || speed > props->f_max) {
+			LOG_ERR("Speed range error %d", speed);
+			return -ENOTSUP;
+		}
+		clk_div = convert_freq_to_divider(speed);
+		MXC_SDHC_Set_Clock_Config(clk_div);
+	}
+
+	if (ios->power_mode == SDHC_POWER_OFF) {
+		MXC_SDHC_PowerDown();
+	} else {
+		MXC_SDHC_PowerUp();
```

Comment:
Is this function a no-op if the card is already powered up? If so, I think this is fine- just want to make sure